### PR TITLE
Add changes to support tf12

### PR DESCRIPTION
**Pull request by evkuzin**
1 - change google_compute_health_check to support all http params
2 - delete some unnecessary []
regarding #3 #10

### CODE REVIEW

**Comment by evkuzin** [comment]
@morgante I've check examples and they depend on another module which still tf11. So what will be the best way - update GoogleCloudPlatform/managed-instance-group/google and then refactor these examples or what?

**Comment by aaron-lane** [comment]
@evkuzin please update the examples to use `terraform-google-modules/vm/google//modules/mig`.

**Comment by evkuzin** [comment]
Update with examples. But examples also use module with external LB which is still tf11. So I erased it from examples.

File: examples/simple/main.tf [hunk 1]
```diff
@@ -26,35 +26,38 @@ variable zone {
...
-  region = "${var.region}"
+variable "service_port" {
+  default = "80"
+}
+
+variable "service_account" {
+  default = "test"
 }
 
-module "gce-lb-fr" {
```
Comment (aaron-lane) [inline examples/simple/main.tf]: The external load balancer should be maintained in the example by replacing `GoogleCloudPlatform/lb/google` with `terraform-google-modules/lb/google`, `version = "~> 2.0"`.

File: main.tf [hunk 2]
```diff
@@ -46,42 +46,58 @@ resource "google_compute_region_backend_service" "default" {
   protocol         = "${var.ip_protocol}"
   timeout_sec      = 10
   session_affinity = "${var.session_affinity}"
-  backend          = ["${var.backends}"]
+  dynamic "backend" {
+    iterator = group
```
Comment (aaron-lane) [inline main.tf]: This should be named `backend` due to the variable being iterated.

File: main.tf [hunk 3]
```diff
@@ -46,42 +46,58 @@ resource "google_compute_region_backend_service" "default" {
...
   }
 }
 
 resource "google_compute_health_check" "http" {
-  count = "${var.http_health_check ? 1 : 0}"
+  count = "${var.health_check["type"] == "http" ? 1 : 0}"
   project = "${var.project}"
   name    = "${var.name}-hc"
-
+  check_interval_sec = "${var.health_check["check_interval_sec"]}" # - (Optional) How often (in seconds) to send a health check. The default value is 5 seconds.
```
Comment (aaron-lane) [inline main.tf]: `check_interval`, `healthy_threshold`, `timeout_sec`, and `unhealthly_threshold`, should be configured on both `google_compute_health_check` resources.

File: variables.tf [hunk 4]
```diff
@@ -58,9 +58,26 @@ variable ports {
...
+  type = "map"
+  default = {
+    type = "tcp"
+    check_interval_sec = 5
+    healthy_threshold = 2
+    timeout_sec = 5
+    unhealthy_threshold = 2
+  }
+}
 variable http_health_check {
```
Comment (aaron-lane) [inline variables.tf]: This variable has been made obsolete by `health_check` so it should be removed. Additionally, we will need a migration guide which explains the change in the variable interface. terraform-google-project-factory and terraform-google-kubernetes-engine both have guides in their `docs` directories to reference.

File: main.tf (hunk 3 above)
Comment (aaron-lane) [inline main.tf]: These comments should be moved to the variable description.

File: variables.tf [hunk 5]
```diff
@@ -58,9 +58,26 @@ variable ports {
   type        = "list"
```
Comment (aaron-lane) [inline variables.tf]: All variables should be typed with the new type system.

File: variables.tf [hunk 6]
```diff
@@ -58,9 +58,26 @@ variable ports {
...
+  default  = {
+    # host = "${var.http_health_check.host != null ? http_health_check.port : null}" # - (Optional) The value of the host header in the HTTP health check request. If left empty (default value), the public IP on behalf of which this health check is performed will be used.
+    request_path = "/" #  - (Optional) The request path of the HTTP health check request. The default value is /.
+    host = "" # - (Optional) The value of the host header in the HTTP health check request. If left empty (default value), the public IP on behalf of which this health check is performed will be used.
+    response = "" #  - (Optional) The bytes to match against the beginning of the response data. If left empty (the default value), any response will indicate health. The response data can only be ASCII.
+    proxy_header = "NONE" #  - (Optional) Specifies the type of proxy header to append before sending data to the backend, either NONE or PROXY_V1. The default is NONE.
+  }
 }
 
 variable health_port {
```
Comment (aaron-lane) [inline variables.tf]: This variable should be merged in to `health_check`.

File: main.tf [hunk 7]
```diff
@@ -46,42 +46,58 @@ resource "google_compute_region_backend_service" "default" {
...
+  healthy_threshold = "${var.health_check["healthy_threshold"]}" # - (Optional) A so-far unhealthy instance will be marked healthy after this many consecutive successes. The default value is 2.
+  timeout_sec = "${var.health_check["timeout_sec"]}" # - (Optional) How long (in seconds) to wait before claiming failure. The default value is 5 seconds. It is invalid for timeoutSec to have greater value than checkIntervalSec.
+  unhealthy_threshold = "${var.health_check["unhealthy_threshold"]}" # - (Optional) A so-far healthy instance will be marked unhealthy after this many consecutive failures. The default value is 2.
   http_health_check {
     port = "${var.health_port}"
+    # host = "${var.http_health_check.host != null ? http_health_check.port : null}" # - (Optional) The value of the host header in the HTTP health check request. If left empty (default value), the public IP on behalf of which this health check is performed will be used.
```
Comment (aaron-lane) [inline main.tf]: Commented-out code should be removed.

File: main.tf [hunk 8]
```diff
@@ -46,42 +46,58 @@ resource "google_compute_region_backend_service" "default" {
...
+    request_path = "${var.http_health_check["request_path"]}" #  - (Optional) The request path of the HTTP health check request. The default value is /.
+    host = "${var.http_health_check["host"]}" # - (Optional) The value of the host header in the HTTP health check request. If left empty (default value), the public IP on behalf of which this health check is performed will be used.
+    response = "${var.http_health_check["response"]}" #  - (Optional) The bytes to match against the beginning of the response data. If left empty (the default value), any response will indicate health. The response data can only be ASCII.
```
Comment (aaron-lane) [inline main.tf]: `response` `port_name`, and `proxy_header` should be configured on both `google_compute_health_check` resources.

File: variables.tf [hunk 9]
```diff
@@ -58,9 +58,26 @@ variable ports {
   type        = "list"
 }
 
+variable health_check {
+  type = "map"
```
Comment (aaron-lane) [inline variables.tf]: This should be an `object` with the supported keys included.

File: examples/simple/mig.tf [hunk 10]
```diff
@@ -14,66 +14,32 @@
...
-  vars {
-    PROXY_PATH = ""
-  }
-}
-
-data "template_file" "group3-startup-script" {
-  template = "${file("${format("%s/gceme.sh.tpl", path.module)}")}"
-
-  vars {
+module "instance_template" {
```
[the remark this answers is not in the document]
Comment (evkuzin) [reply examples/simple/mig.tf]: Replace this module with `data "template_file"`

File: variables.tf [hunk 11]
```diff
@@ -55,26 +55,44 @@ variable session_affinity {
...
+    response            = string
+    proxy_header        = string
+    health_port         = number
+  })
+  default = null
 }
-
-variable health_port {
-  description = "Port to perform health checks on."
+variable "http_health_check" {
```
Comment (aaron-lane) [inline variables.tf]: This variable needs to be removed; it is redundant since `health_check` has `type`.

**Comment by aaron-lane** [comment]
@omazin please take this on.

**Pull request closed without merging**